src/topology/metric_space/isometry.lean

```diff
@@ -216,6 +218,39 @@ begin
   rw [dist_eq_norm, dist_eq_norm, ← algebra.map_sub, norm_algebra_map_eq],
 end
 
+lemma lipschitz_with.injective_add_isometry [metric_space α] [normed_group β] {f : α → β}
+  {c : nnreal} (hf : lipschitz_with c f) (hc : c < 1) {e : α → β} (he : isometry e) :
+  injective (λ x, f x + e x) :=
+begin
+  assume x y hxy,
+  have A : dist (f x) (f y) ≤ c * dist x y := hf x y,
+  have B : dist (f x) (f y) = dist (e x) (e y),
+  { rw [dist_eq_norm, dist_eq_norm, norm_sub_rev],
+    apply congr_arg,
+    rw [← sub_eq_iff_eq_add] at hxy,
+    simp only [hxy.symm],
+    abel },
+  have : 0 ≤ ((c:ℝ) - 1) * dist x y,
+    by simpa only [sub_mul, one_mul, sub_nonneg, B, he.dist_eq] using A,
+  exact dist_le_zero.1 (nonpos_of_mul_nonneg_right this (sub_lt_zero.2 hc))
+end
+
+lemma lipschitz_with.inj_on_add_isometry [metric_space α] [normed_group β] {f : α → β}
+  {s : set α} {c : nnreal} (hf : lipschitz_with c (restrict f s)) (hc : c < 1)
+  {e : α → β} (he : isometry e) :
+  inj_on (λ x, f x + e x) s :=
+inj_on_iff_injective.2 $ hf.injective_add_isometry hc (he.restrict s)
+
+lemma contracting_with.injective_add_id [normed_group α] {f : α → α}
+  {c : nnreal} (hf : contracting_with c f) :
+  injective (λ x, f x + x) :=
+hf.2.injective_add_isometry hf.1 isometry_id
+
+lemma lipschitz_with.inj_on_add_id [normed_group α] {f : α → α} {s : set α}
+  {c : nnreal} (hf : lipschitz_with c (restrict f s)) (hc : c < 1) :
+  inj_on (λ x, f x + x) s :=
```

Comment:
You have some lemmas with assumptions `lipschitz_with` and some lemmas with `contracting_with`. It would make more sense to express everything in terms of `contracting_with`, as these lemmas are really about contracting maps.

Reply:
`contracting_with f` is defined for `f : α → α` because this is the only case we need for the Banach fixed point theorem. Possibly I should redefine `contracting_with`.

Reply:
Yes, there is no reason why `contracting_with` should be restricted to functions from one type to itself.

Reply:
It came to my mind that actually these lemmas are about the following two facts:

1. If `∀ x y, C * dist x y ≤ dist (f x) (f y)`, `lipschitz_with c f`, and `c < C`, then `∀ x y, (C - c) * dist x y ≤ dist (f x) (f y)`.
2. If `∀ x y, C * dist x y ≤ dist (f x) (f y)`, `0 < C`, then `f` is injective.

Should we introduce `∀ x y, C * dist x y ≤ dist (f x) (f y)` as a named predicate on functions? If yes, how would you name it?

Reply:
Maybe `expanding_with`?

Reply:
No, `expanding_with` is not good, it seems to imply `C>1`. `antilipschitz with`?